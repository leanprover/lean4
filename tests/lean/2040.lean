/-! Unification across `calc` steps -/

-- Tests were written when `^` was still using `binop%`
macro_rules | `($x ^ $y) => `(binop% HPow.hPow $x $y)

example (n : Nat) (a : Int) : a = 22 :=
  calc
    a = 2 ^ n := sorry -- error
    _ = (22 : Int) := sorry

example (n : Nat) (a : Int) : a = 22 :=
  calc
    a = (37 : Int) := sorry
    _ = 2 ^ n := sorry -- should be same error as above
    _ = (22 : Int) := sorry

example (n : Nat) (a : Int) : a = (2 : Int) ^ n :=
  calc
    a = (37 : Int) := sorry
    _ = 2 ^ n := sorry -- should be same error as above

example (n : Nat) (h : n = 42) : 42 = (n : Int) :=
  calc
    (42 : Int) = 42 := rfl
    _ = n := h ▸ rfl
      --^ coe needs to be inserted here

example (n : Nat) (h : n = 42) : 42 = (n : Int) :=
  calc
    (42 : Int) = 42 := rfl -- TODO: type of 42 should match goal (i.e., `Int`)
    _ = n := h ▸ rfl
      --^ coe needs to be inserted here

example (n : Nat) (h : n = 42) : 42 = (n : Int) :=
  calc
    (_ : Int) = 42 := rfl -- TODO: type of 42 should match goal (i.e., `Int`)
    _ = n := h ▸ rfl
      --^ coe needs to be inserted here

example := calc 1 = 1 := rfl
example := by calc 1 = 1 := rfl
