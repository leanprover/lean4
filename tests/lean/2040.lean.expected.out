2040.lean:8:8-8:13: error: failed to synthesize instance
  HPow Nat Nat Int
2040.lean:14:8-14:13: error: failed to synthesize instance
  HPow Nat Nat Int
2040.lean:20:8-20:13: error: failed to synthesize instance
  HPow Nat Nat Int
2040.lean:18:2-20:22: error: type mismatch
  trans (sorryAx (a = 37)) (sorryAx (37 = 2 ^ n))
has type
  a = @OfNat.ofNat Nat 2 (instOfNatNat 2) ^ n : Prop
but is expected to have type
  a = @OfNat.ofNat Int 2 instOfNatInt ^ n : Prop
